### ISSUE TITLE
/var/vcap/data on diego-cell needs to be a separate volume with quota support

### ISSUE THREAD

**Issue by jandubois** [cloudfoundry-incubator/kubecf#441]
# The problem

garden and grootfs both assume that they together have exclusive control over the disk space mounted at `/var/vcap/data`. The property `grootfs.reserved_space_for_other_jobs_in_mb` (default: 15360) can be used to carve out space for other jobs.

Currently an `emptyDir` volume is mounted to various parts of the filesystem. I don't understand what purpose this serves, as the disk space will still be allocated on the local node storage. Yes, it survives a pod restart, but I'm not even sure this is desirable.

The real issue though is that now grootfs is going to allocate sparse files as backing store for the privileged and unprivileged stores that claim all the size except for 15360MB of the local node storage, and formats those sparse files with XFS.

There is more to this, including the way garden calculates available disk space for apps, and how it does garbage collection for the layer cache. That code assumes that `/var/vcap/data/garden/depot` is on the same disk as `/var/vcap/data` and the grootfs stores.

# Potential fix

Use a PVC and mount it to `/var/vcap/data`. This still requires that the storageclass for this PVC implements quotas correctly. The NFS test setup used during development typically does not, so would still allocate huge stores.

This is the mechanism SCF uses to move the grootfs away from the local node disk. For some reason we found it necessary to wipe this PVC every time the container restarts. I don't know why this was necessary, but the current `emptyDir` volume used by kubecf would keep data between restarts and only be wiped when the pod is deleted.

# Alternative

Assuming that we can require worker nodes to have sufficient disk space for the grootfs, we might be able to set a `limit.disk` for the pod to get a quota on the pod filesystem. That way we could get rid of the volume mount completely.

This would also provide faster storage (just like the current `emptyDir` setup) in case the node uses e.g. SSD drives.

**Comment by jandubois** [comment]
For illustration here you can see the `dh -h` output from inside the `diego-cell`, and the `/var/vcap/data/grootfs/store/unprivileged` claims to have 419G available where as the disk itself only has 360G of free space:

```
/:/var/vcap/jobs/garden# df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay         436G   77G  360G  18% /
tmpfs            64M     0   64M   0% /dev
tmpfs            16G     0   16G   0% /sys/fs/cgroup
/dev/sda4       436G   77G  360G  18% /etc/hosts
shm              64M     0   64M   0% /dev/shm
tmpfs            16G   12K   16G   1% /run/secrets/kubernetes.io/serviceaccount
/dev/loop0      421G  1.6G  419G   1% /var/vcap/data/grootfs/store/unprivileged
/dev/loop1      421G  462M  420G   1% /var/vcap/data/grootfs/store/privileged
```

**Comment by f0rmiga** [comment]
@fargozhu We need to determine if this is a blocker for KubeCF 1.0.

**Comment by f0rmiga** [comment]
Docs for the grootfs disk usage: https://github.com/cloudfoundry/garden-runc-release/blob/91f02154969fc94546d766573c047ad2eb7dff2f/docs/understanding_grootfs_store_disk_usage.md.

**Comment by viovanov** [comment]
@jandubois it seems that using PVCs is making cells unstable. Specifically because a lot of storage classes are based on NFS, some with hard mounts. These cause garden and the cell to missbehave (write/close calls hanging indefinitely).
I think garden was designed with local storage in mind for garden store.

**Comment by viovanov** [comment]
@jandubois  with the disk capacity set as below, the cell reports the expected capacity.
Based on your original comment, it also sounds like we should calculate `grootfs.reserved_space_for_other_jobs_in_mb` at runtime. I'm working on a patch. Did you use some automated testing to confirm the cell was consuming more than it should? 

```
properties:
  diego-cell:
    rep:
      diego:
        executor:
          disk_capacity_mb: 40960
```

**Comment by jandubois** [comment]
@viovanov The `disk_capacity_mb` is a `rep` property, which doesn't limit how much space is allocated by `grootfs init-store`. That size will be calculated based on the size of the filesystem mounted at `/var/vcap/data` unless you explicitly override it with a `--store-size-bytes` commandline option to `grootfs init-store`.

Since this is only allocating a sparse file, this may still be fine as long as you don't use too much space inside. I think part of the problem before was that the sparse file was allocated with a size larger than the partition size because a different PVC was mounted under `/var/vcap/data/grootfs` that was smaller than the filesystem at `/var/vcap/data`.

But really, you are now relying on details of how the sparse file logic is implemented by the PVC **and** how the xfs filesystem is managing disk blocks inside the sparse file, which sounds to me like: "Let's hope this all works out!".

E.g. I don't see any reason why a filesystem should try to re-use deleted blocks as soon as possible and not use up all available disk blocks before re-using. The whole grootfs logic relies on the filesystem reusing disk blocks as much as possible in order to avoid committing more blocks from the sparse filesystem. Except XFS doesn't know that this is a concern.

So in summary, I think overallocating the sparse file is asking for problems, even if it seems to work properly in our specific testing.

I really wish I would never have to think about this again, it is such a mess!

**Comment by viovanov** [comment]
#1264 contains a patch to calculate sparse file size
